src/colmap/estimators/bundle_adjustment.cc

```
@@ -419,21 +422,214 @@ void ParameterizeCameras(const BundleAdjustmentOptions& options,
+  }
+}
+
+void ParameterizeImages(const BundleAdjustmentOptions& options,
+                        const BundleAdjustmentConfig& config,
+                        const std::set<image_t>& image_ids,
+                        Reconstruction& reconstruction,
+                        ceres::Problem& problem) {
+  std::unordered_set<sensor_t> parameterized_sensor_ids;
+  std::unordered_set<frame_t> parameterized_frame_ids;
+  for (const image_t image_id : image_ids) {
+    Image& image = reconstruction.Image(image_id);
+    if (image.HasTrivialFrame()) {
+      // Parameterize cam_from_world (==frame_from_world).
+      if (parameterized_frame_ids.insert(image.FrameId()).second) {
+        Rigid3d& cam_from_world = image.FramePtr()->FrameFromWorld();
+        // CostFunction assumes unit quaternions.
+        cam_from_world.rotation.normalize();
+        if (problem.HasParameterBlock(
+                cam_from_world.rotation.coeffs().data())) {
+          SetQuaternionManifold(&problem,
+                                cam_from_world.rotation.coeffs().data());
+          if (!options.refine_frame_from_world ||
+              config.HasConstantFrameFromWorldPose(image.FrameId())) {
+            problem.SetParameterBlockConstant(
+                cam_from_world.rotation.coeffs().data());
+            problem.SetParameterBlockConstant(
+                cam_from_world.translation.data());
+          }
+        }
+      }
+    } else {
+      // Parameterize sensor_from_rig.
+      const sensor_t sensor_id = image.CameraPtr()->SensorId();
+      if (parameterized_sensor_ids.insert(sensor_id).second) {
+        Rigid3d& sensor_from_rig =
+            image.FramePtr()->RigPtr()->SensorFromRig(sensor_id);
+        // CostFunction assumes unit quaternions.
+        sensor_from_rig.rotation.normalize();
+        if (problem.HasParameterBlock(
+                sensor_from_rig.rotation.coeffs().data())) {
+          SetQuaternionManifold(&problem,
```

nit: Do you think it is makes sense to do the ``HasParameterBlock`` check in the ``SetManifold`` functions the colmap? We can also create a ``SetParameterBlockConstant`` function similarly.

---

I'd like to avoid that so we don't silently swallow errors. In this specific use case, we only want to set the manifold when it was added to the problem. However, in other cases, this behavior is not desirable (imo).

---

Okay. This makes sense to me. We could still add function like SetParameterBlockConstantIfPresent though xD.